winlogbeat/eventlog/eventlog.go

```diff
@@ -35,100 +41,133 @@ type EventLog interface {
-		"@timestamp":            common.Time(r.TimeGenerated),
-		"log_name":              r.EventLogName,
-		"source_name":           r.SourceName,
-		"computer_name":         r.ComputerName,
-		// Use a string to represent this uint64 data because its value can
-		// be outside the range represented by a Java long.
-		"record_number": strconv.FormatUint(r.RecordNumber, 10),
-		"event_id":      r.EventID,
-		"level":         r.Level,
-		"type":          r.API,
-		"count":         1,
+		"type":                  e.API,
+		common.EventMetadataKey: e.EventMetadata,
+		"count":                 1,
+		"@timestamp":            common.Time(e.TimeCreated.SystemTime),
+		"log_name":              e.Channel,
+		"source_name":           e.Provider.Name,
+		"computer_name":         e.Computer,
+		"record_number":         strconv.FormatUint(e.RecordID, 10),
+		"event_id":              e.EventIdentifier.ID,
 	}
 
-	if r.Message != "" {
-		m["message"] = r.Message
-	} else {
-		if len(r.MessageInserts) > 0 {
-			m["message_inserts"] = r.MessageInserts
+	addOptional(m, "provider_guid", e.Provider.GUID)
+	addOptional(m, "version", e.Version)
+	addOptional(m, "level", e.Level)
+	addOptional(m, "task", e.Task)
+	addOptional(m, "opcode", e.Opcode)
+	addOptional(m, "keywords", e.Keywords)
+	addOptional(m, "message", sys.RemoveWindowsLineEndings(e.Message))
+	addOptional(m, "message_error", e.RenderErr)
+
+	// Correlation
+	addOptional(m, "activity_id", e.Correlation.ActivityID)
+	addOptional(m, "related_activity_id", e.Correlation.RelatedActivityID)
+
+	// Execution
+	addOptional(m, "process_id", e.Execution.ProcessID)
+	addOptional(m, "thread_id", e.Execution.ThreadID)
+	addOptional(m, "processor_id", e.Execution.ProcessorID)
+	addOptional(m, "session_id", e.Execution.SessionID)
+	addOptional(m, "kernel_time", e.Execution.KernelTime)
+	addOptional(m, "user_time", e.Execution.UserTime)
+	addOptional(m, "processor_time", e.Execution.ProcessorTime)
+
+	if e.User.Identifier != "" {
+		user := common.MapStr{
+			"identifier": e.User.Identifier,
 		}
+		m["user"] = user
 
-		if r.MessageErr != nil {
-			m["message_error"] = r.MessageErr.Error()
-		}
+		addOptional(user, "name", e.User.Name)
+		addOptional(user, "domain", e.User.Domain)
+		addOptional(user, "type", e.User.Type.String())
+	}
+
+	addPairs(m, "event_data", e.EventData.Pairs)
+	userData := addPairs(m, "user_data", e.UserData.Pairs)
+	addOptional(userData, "xml_name", e.UserData.Name.Local)
+
+	return m
+}
+
+// addOptional adds a key and value to the given MapStr if the value is not the
+// zero value for the type of v. It is safe to call the function with a nil
+// MapStr.
+func addOptional(m common.MapStr, key string, v interface{}) {
+	if m != nil && !isZero(v) {
+		m[key] = v
 	}
+}
 
-	if r.Category != "" {
-		m["category"] = r.Category
+// addPairs adds a new dictionary to the given MapStr. The key/value pairs are
+// added to the new dictionary. If any keys are duplicates, the first key/value
```

Comment:
We could use something like that also in Metricbeat. Potentially worth to abstract out.